File: 04-decision-tree/decision_tree.py
import pandas as pd
from sklearn.model_selection import train_test_split
from sklearn.metrics import confusion_matrix, roc_curve, auc, accuracy_score
from sklearn.tree import DecisionTreeClassifier, export_graphviz
from sklearn.preprocessing import LabelEncoder
import matplotlib.pyplot as plt
import graphviz

# Read dataset
titanic = pd.read_csv("titanic.csv").dropna(subset=["Survived"])
print(titanic.describe())
print(titanic.columns)

# Factorize columns
label_encoder = LabelEncoder()
titanic["Sex"] = label_encoder.fit_transform(titanic["Sex"])
titanic["Survived"] = label_encoder.fit_transform(titanic["Survived"])
titanic["Embarked"] = label_encoder.fit_transform(titanic["Embarked"])

# Divide into train and test set
x_train, x_test, y_train, y_test = train_test_split(titanic[["Pclass", "Sex", "Age", "SibSp", "Parch", "Fare", "Embarked"]], titanic["Survived"], test_size=0.2, random_state=0)

# Train decision tree classifier
clf = DecisionTreeClassifier(random_state=42).fit(x_train, y_train)

# Plot tree
dot_data = export_graphviz(clf, out_file=None, feature_names=["Pclass", "Sex", "Age", "SibSp", "Parch", "Fare", "Embarked"], filled=True, rounded=True, special_characters=True)
graph = graphviz.Source(dot_data)
graph.render("tree")

# Pruning
cf_pruned = DecisionTreeClassifier(random_state=42, max_depth=5, min_samples_split=10, min_samples_leaf=5).fit(x_train, y_train)
dot_data = export_graphviz(cf_pruned, out_file=None, feature_names=["Pclass", "Sex", "Age", "SibSp", "Parch", "Fare", "Embarked"], filled=True, rounded=True, special_characters=True)
graph = graphviz.Source(dot_data)
graph.render("tree_pruned")

# Evaluate model
y_pred = clf.predict(x_test)
cm = confusion_matrix(y_test, y_pred)
accuracy = accuracy_score(y_test, y_pred) * 100
print(f"Confusion Matrix: {cm}")
print(f"Accuracy: {accuracy}")

# ROC curve
y_probs = clf.predict_proba(x_test)[:, 1]
fpr, tpr, thresholds = roc_curve(y_test, y_probs)
roc_auc = auc(fpr, tpr)
plt.figure(figsize=(10, 6))
plt.plot(fpr, tpr, color="darkorange", lw=2, label="ROC curve (area = {:.2f})".format(roc_auc))
plt.plot([0, 1], [0, 1], color="navy", lw=2, linestyle="--")
plt.xlabel("False Positive Rate")
plt.ylabel("True Positive Rate")
plt.title("Receiver Operating Characteristic (ROC) Curve")
plt.legend(loc="lower right")
plt.show()

File: 04-decision-tree/decision_tree.R
# Install packages
install.packages("partykit")
install.packages("caret")
install.packages("pROC")
install.packages('rattle')
install.packages('rpart.plot')
install.packages('RColorBrewer')

# Read dataset
titanic <- read.csv("titanic.csv")
summary(titanic)
names(titanic)

# Factorize columns
titanic$Sex <- as.factor(titanic$Sex)
titanic$Survived <- as.factor(titanic$Survived)
titanic$Embarked <- as.factor(titanic$Embarked)

# Sample dataset
pd <- sample(2, nrow(titanic), replace = TRUE, prob = c(0.8, 0.2))
pd

# Divide into train and test set
train_set <- titanic[pd == 1,]
test_set <- titanic[pd == 2,]

# Remove nulls
train_set <- train_set[complete.cases(train_set$Survived),]
test_set <- test_set[complete.cases(test_set$Survived),]

# Plot tree
library("partykit")
tree <- ctree(formula = Survived ~ Pclass + Sex + Age + SibSp + Parch + Fare + Embarked, data= train_set)
plot(tree)

# Pruning
tree <- ctree(formula = Survived ~ Pclass + Sex + Age + SibSp + Parch + Fare + Embarked, data = train_set,control = ctree_control(mincriterion = 0.99, minsplit = 500))
plot(tree)
pred <- predict(tree, test_set, type = "prob")
pred

# Ensure that both vectors have the same length
pred <- predict(tree, test_set)
pred <- pred[complete.cases(test_set$Survived)]
survived <- test_set$Survived[complete.cases(test_set$Survived)]

library("caret")
confusionMatrix(pred, survived)

# Predict on the entire test set including missing values
pred_prob <- predict(tree, test_set, type = "prob")

# Extract the predicted probabilities for the positive class
pred_probabilities <- as.numeric(pred_prob[, "1"])

# Compute the ROC curve
library("pROC")
roc_curve <- roc(test_set$Survived, pred_probabilities)

# Plot the ROC curve
plot(roc_curve)

library(rpart)
fit <- rpart(Survived ~ Pclass + Sex + Age + SibSp + Parch + Fare + Embarked, data = titanic, method = "class")
plot(fit)
text(fit)

library(rattle)
library(rpart.plot)
library(RColorBrewer)
fancyRpartPlot(fit)
prediction <- predict(fit, titanic, type = "class")
prediction